docs/building-modules/msg-services.md

```diff
@@ -0,0 +1,99 @@
+<!--
+order: 4
+-->
+# `Msg` Services
+
+A `Msg` Service processes [`message`s](./messages-and-queries.md#messages). `Msg` Services are specific to the module in which they are defined, and only process `message`s defined within the said module. They are called from `baseapp` during [`DeliverTx`](../core/baseapp.md#delivertx). {synopsis}
+
+## Pre-requisite Readings
+
+- [Module Manager](./module-manager.md) {prereq}
+- [Messages and Queries](./messages-and-queries.md) {prereq}
+
+## Implementation of a module `Msg` service
+
+All `Msg` processing is done by a `MsgServer`](messages-and-queries.md#msg-services) protobuf service. Each module should define a `Msg` service, which will be responsible for request and response serialization. 
+
+As further described in [ADR 031](../architecture/adr-031-msg-service.md), this approach has the advantages of clearly specifying return types and generating server and client code.
+
+When possible, the existing module's [`Keeper`](keeper.md) should implement `MsgServer`, else a `msgServer` struct that embeds the `Keeper` can be created, typically in `./keeper/msg_server.go`:
+
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc1/x/bank/keeper/msg_server.go#L14-L16
+
+At the moment, we use the SDK's [existing routing mechanism](#handler-type) to route a `Msg` to the appropriate `MsgServer` method. But this is intended to involve to satisfy the use cases discussed in [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093)
```

Comment:
wait, I think this is incorrect, we updated the routing of new Msgs to use `msgServiceRouter`: https://github.com/cosmos/cosmos-sdk/blob/4420fe2d52923613bb0588b8141ce80e13b1418d/baseapp/baseapp.go#L705-L710